polkadot/node/core/candidate-validation/src/lib.rs

```diff
@@ -298,6 +308,247 @@ async fn run<Context>(
+	// It's ok to stil prepare PVFs in other cases, but better to request only future ones.
+	let Ok(Ok(authorities)) = util::request_authorities(relay_parent, sender).await.await else {
+		gum::warn!(
+			target: LOG_TARGET,
+			?relay_parent,
+			"cannot fetch authorities from runtime API",
+		);
+		return false
+	};
+
+	// We need to exclude at least current session authority from the previous request
+	let Ok(Ok(Some(session_info))) =
+		util::request_session_info(relay_parent, session_index, sender).await.await
+	else {
+		gum::warn!(
+			target: LOG_TARGET,
+			?relay_parent,
+			"cannot fetch session info from runtime API",
+		);
+		return false
+	};
+
+	let is_past_present_or_future_authority = authorities
+		.iter()
+		.any(|v| keystore.has_keys(&[(v.to_raw_vec(), AuthorityDiscoveryId::ID)]));
+
+	let is_present_authority = session_info
+		.discovery_keys
+		.iter()
+		.any(|v| keystore.has_keys(&[(v.to_raw_vec(), AuthorityDiscoveryId::ID)]));
+
+	// There is still a chance to be a previous session authority, but this extra work does not
+	// affect the finalization.
+	is_past_present_or_future_authority && !is_present_authority
+}
+
+// Sends PVF with unknown code hashes to the validation host returning the list of code hashes sent.
+async fn prepare_pvfs_for_backed_candidates<Sender>(
+	sender: &mut Sender,
+	mut validation_backend: impl ValidationBackend,
+	relay_parent: Hash,
+	already_prepared: &HashSet<ValidationCodeHash>,
+	per_block_limit: usize,
+) -> Option<Vec<ValidationCodeHash>>
+where
+	Sender: SubsystemSender<RuntimeApiMessage>,
+{
+	let Ok(Ok(events)) = util::request_candidate_events(relay_parent, sender).await.await else {
+		gum::warn!(
+			target: LOG_TARGET,
+			?relay_parent,
+			"cannot fetch candidate events from runtime API",
+		);
+		return None
+	};
+	let code_hashes = events
+		.into_iter()
+		.filter_map(|e| match e {
+			CandidateEvent::CandidateBacked(receipt, ..) => {
+				let h = receipt.descriptor.validation_code_hash;
+				if already_prepared.contains(&h) {
+					None
+				} else {
+					Some(h)
+				}
+			},
+			_ => None,
+		})
+		.take(per_block_limit)
+		.collect::<Vec<_>>();
+
+	let Ok(executor_params) = util::executor_params_at_relay_parent(relay_parent, sender).await
+	else {
+		gum::warn!(
+			target: LOG_TARGET,
+			?relay_parent,
+			"cannot fetch executor params for the session",
+		);
+		return None
+	};
+	let timeout = pvf_prep_timeout(&executor_params, PvfPrepKind::Prepare);
+
+	let mut active_pvfs = vec![];
+	let mut processed_code_hashes = vec![];
+	for code_hash in code_hashes {
+		let Ok(Ok(Some(validation_code))) =
+			util::request_validation_code_by_hash(relay_parent, code_hash, sender)
+				.await
+				.await
+		else {
+			gum::warn!(
+				target: LOG_TARGET,
+				?relay_parent,
+				?code_hash,
+				"cannot fetch validation code hash from runtime API",
+			);
+			continue;
+		};
+
+		let pvf = match sp_maybe_compressed_blob::decompress(
```

Comment:
I think this just keeps decompressing needlessly if we are a validator intermittently: at session 1, then at session 2 we are no longer, but we are again at session 3. 

All that was prepared for session 1 will be reset as soon as we go into session 2 (`already_prepared_code_hashes` is cleared). Is there a better way to check? Like query the PVF subsystem if it has it already ? Also I think the query needs to take into account the executor params which we are preparing with not just code hash.

Reply:
let's handle it a new PR.

Reply:
> let's handle it a new PR.

Or maybe will not be even needed: https://github.com/paritytech/polkadot-sdk/issues/5071#issuecomment-2241014329

Reply:
I think we still need to polish it. #5071 ticket talks about the other path, where we actually do candidate validation, here we just try to compile artifacts for blocks backed on chain. Given that, I expect decompression should be fast, the only thing concerning here is that we do it for all blocks.